src/validation/rules/optional/ProhibitDeprecatedFieldsRule.js

```diff
@@ -0,0 +1,52 @@
+// @flow strict
+
+import { GraphQLError } from '../../../error/GraphQLError';
+
+import { type EnumValueNode, type FieldNode } from '../../../language/ast';
+import { type ASTVisitor } from '../../../language/visitor';
+
+import { getNamedType } from '../../../type/definition';
+
+import { type ValidationContext } from '../../ValidationContext';
+
+/**
+ * Prohibit deprecated fields
+ *
+ * A GraphQL document is only valid if all selected fields and all used enum
+ * values have not been deprecated.
+ *
+ * Note: This rule is optional and is not part of the Validation section of the
+ * GraphQL Specification.
+ */
+export function ProhibitDeprecatedFieldsRule(
+  context: ValidationContext,
+): ASTVisitor {
+  return {
+    Field(node: FieldNode) {
+      const fieldDef = context.getFieldDef();
+      const parentType = context.getParentType();
+      if (parentType && fieldDef && fieldDef.deprecationReason != null) {
```

Comment:
I modified the original logic here to account for the fact that a deprecated field will always have a deprecation reason (since `isDeprecated` is derived from the truthiness of the `deprecationReason` value). The original function had code that was effectively unreachable because it tried to take into account a scenario where `isDeprecated` is true but no `deprecationReason` is defined.

I'd rather test for `isDeprecated` here, but then we'd have to cast `fieldDef.deprecationReason` to `any` and again to `string`, and this seemed like the lesser evil.

Ideally, GraphQLField would be a union type like `{ isDeprecated: true, deprecationReason: string } | { isDeprecated: false, deprecationReason: null | void }` -- maybe that's something we could change?

Reply:
I actually think `isDeprecated` itself was a mistake and we should just use `deprecationReason` everywhere.
BTW. You can create and issue for that mark `isDeprecated` as deprecated :)
and update all our code to use `deprecationReason` instead.